Client/game_sa/gamesa_init.h

```diff
@@ -104,3 +104,45 @@ void MemOrFast(U ptr, const T value)
 
 bool GetDebugIdEnabled(uint uiDebugId);
 void LogEvent(uint uiDebugId, const char* szType, const char* szContext, const char* szBody, uint uiAddReportLogId = 0);
+
+struct __THISCALL{};
+struct __CDECL{};
+struct __STDCALL{};
+struct __FASTCALL{};
+struct __VECTORCALL{};
+
+template <typename... Args>
+struct GTAFuncSignature
+{
+    std::tuple<Args...> args;
+    GTAFuncSignature(Args... a) : args(std::forward<Args>(a)...) {}
+};
+
+template <typename... Args>
+GTAFuncSignature<Args...> PrepareSignature(Args... args)
+{
+    return GTAFuncSignature<Args...>(std::forward<Args>(args)...);
+}
+
+template <typename ReturnType, typename Func, typename Tuple, std::size_t... I>
+ReturnType apply_helper(Func&& f, Tuple&& t, std::index_sequence<I...>)
+{
+    return f(std::get<I>(std::forward<Tuple>(t))...);
+}
+
+template <typename ReturnType, typename CallingConvention, typename Func, typename... Args>
+ReturnType CallGTAFunction(Func function, GTAFuncSignature<Args...>& sig)
+{
+    if constexpr (std::is_same_v<CallingConvention, __THISCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__thiscall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __CDECL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__cdecl*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __STDCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__stdcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __FASTCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__fastcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else if constexpr (std::is_same_v<CallingConvention, __VECTORCALL>)
+        return apply_helper<ReturnType>(reinterpret_cast<ReturnType(__vectorcall*)(Args...)>(function), sig.args, std::index_sequence_for<Args...>{});
+    else
+        static_assert(sizeof(char) == 0, "Invalid calling convention specified in CallGTAFunction");
```

Comment:
This makes your function ill-formed. Please, use SFINAE for this.

Reply:
The point of that is to throw an error, not discard the result. The same thing happens in ArgumentParser.
How does it make the function ill formed?

Reply:
> The point of that is to throw an error, not discard the result. The same thing happens in ArgumentParser. How does it make the function ill formed ~beside that it will error if invalid specialization is provided which is the point of it~?

See the standard. The extraction from it:
```
The program is ill-formed, no diagnostic required, if:
- no valid specialization can be generated for a template or a substatement of a constexpr if statement within a template and the template is not instantiated, or
...
```

Reply:
\> See the standard. The extraction from it:
> 
> ```
> The program is ill-formed, no diagnostic required, if:
> - no valid specialization can be generated for a template or a substatement of a constexpr if statement within a template and the template is not instantiated, or
> ...
> ```

The point of this code is to stop the compilation process when a specialization couldnt be generated. Check https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2593r1.html

Reply:
Do you know that this is just a proposal? According to the standard this is still the ill-formed program.

Reply:
This is not only a proposal, it was implemented in current versions of GCC and Clang.
The standard requires that a valid specialization has to exist for each if constexpr branch, not only for the entire template.

Reply:
> This makes your function ill-formed. Please, use SFINAE for this.

You mean something like this?
```cpp
template <typename ReturnType, typename Func, typename... Args>
typename std::enable_if<false, ReturnType>::type
CallGTAFunction(Func, Args...)
{
    static_assert(sizeof(Func) == 0, "Invalid calling convention specified in CallGTAFunction");
}
```

Reply:
> This is not only a proposal, it was implemented in current versions of GCC and Clang. The standard requires that a valid specialization has to exist for each if constexpr branch, not only for the entire template.
> 
> `sizeof` is used instead of false for some previous compiler bugs where you had to trick the compiler into always erroring

The fact that it may work on some compilers doesn't cancel the deviation from the standard. The code must work on all possible compilers.

Reply:
> > This makes your function ill-formed. Please, use SFINAE for this.
> 
> You mean something like this?
> 
> ```c++
> template <typename ReturnType, typename CallingConvention, typename Func, typename... Args>
> typename std::enable_if<std::is_same<CallingConvention, __THISCALL>::value, ReturnType>::type CallGTAFunction(Func function, Args... params)
> {
>     return reinterpret_cast<ReturnType(__thiscall*)(Args...)>(function)(params...);
> }
> ```

Yes

Reply:
> The fact that it may work on some compilers doesn't cancel the deviation from the standard. The code must work on all possible compilers.

This is now a standard, not that it may or may not work on some compilers.
If compilers dont follow a standard then we shouldnt pay much attention to them either way.
This will work on all compilers that already support `static_assert(false,...)` statement. Very old versions of gcc and clang might not support that, but then they also dont support advanced templating used in ArgumentParser.
Since ArgumentParser isnt causing any known issues, this code will work on all computers as expected.

Reply:
> > The fact that it may work on some compilers doesn't cancel the deviation from the standard. The code must work on all possible compilers.
> 
> This is now a standard, not that it may or may not work on some compilers. If compilers dont follow a standard then we shouldnt pay much attention to them either way. This will work on all compilers that already support `static_assert(false,...)` statement. Very old versions of gcc and clang might not support that, but then they also dont support advanced templating used in ArgumentParser. Since ArgumentParser isnt causing any known issues, this code will work on all computers as expected.

Okay. Could you provide a proof of including this proposal into the standard?

Reply:
> Okay. Could you provide a proof of including this proposal into the standard?

https://github.com/cplusplus/papers/issues/1251

Reply:
> > Okay. Could you provide a proof of including this proposal into the standard?
> 
> [cplusplus/papers#1251](https://github.com/cplusplus/papers/issues/1251)

It's only C++23. That means that static_asserts are prohibited in templates until MTA get C++23 support. And that's the end of this discussion.

Reply:
> It's only C++23. That means that static_asserts are prohibited in templates until MTA get C++23 support. And that's the end of this discussion.

In either case using SFINAE for this might yield undesirable results

Reply:
Done